hpx/lcos/async.hpp

```diff
@@ -65,8 +90,18 @@ namespace hpx { namespace detail
             static_assert(is_valid::value,
                 "The action to invoke is not supported by the target");
 
-            return hpx::detail::async_impl<Action>(launch_policy, c.get_id(),
-                std::forward<Ts>(ts)...);
+            // invoke directly if client is ready
+            if (c.is_ready())
```

Comment:
It's an optimization which avoids creating a `future<future<>>` for the sole purpose of unwrapping it right away. So yes, we could simply call `.then()` unconditionally if we don't care about this overhead.